### ISSUE TITLE
Benchmark the usefulness of maintaining support for slices

### ISSUE THREAD

**Issue by vdumoulin** [mila-udem/fuel#164]
We'd like to know what is it exactly we're gaining by using slices as opposed to lists of indices for dataset requests.

We should benchmark it and decide whether or not it's worth supporting it.

See [this discussion](https://github.com/mila-udem/fuel/pull/144#discussion-diff-32262341) for more details.

**Comment by nouiz** [comment]
In Theano, doing slicing vs advanced indexing make a significant difference
as this remove a copy.

If when you do slicing you force a copy, maybe the difference isn't as
significant.

**Comment by vdumoulin** [comment]
In our case we're talking about slicing in numpy arrays of HDF5 datasets, but that's a useful thing to know, thanks!

**Comment by dwf** [comment]
In theory there might be NumPy arrays in memory, too, right?

**Comment by nouiz** [comment]
Yes. So it could depend of the size and how it was compressed and stored on
disk and many internal detail.

So timming probably won't be easy. Maybe you could email there mailing list?

**Comment by vdumoulin** [comment]
@dwf Absolutely, I meant to write "numpy arrays **or** HDF5 datasets" but failed miserably.

**Comment by fvisin** [comment]
I did some benchmarks on MNIST (HDF5 file) with the LeNet network.
Here are the results:

Slice | In memory |  Batch size | Time for 2 epochs
----- | --------- | ---------- | -----------------
Yes | Yes | 32 | 712.957221508
No | Yes | 32 | 715.286180735
Yes | No | 32 | 716.082897425
No | No | 32 | 717.730975866
Yes | Yes | 1024 | 711.972523451
No | Yes | 1024 | 713.280206442
Yes | No | 1024 | 715.503113031
No | No | 1024 | 716.049200296

**Comment by dwf** [comment]
Did you turn off the `load_in_memory` option?

**Comment by fvisin** [comment]
It was off only in the experiments where `In memory` is "No".

**Comment by fvisin** [comment]
Here are the results without `default_transformer`:

Slice | In memory |  Batch size | Time for 2 epochs (CPU)
----- | --------- | ---------- | -----------------
Yes | Yes | 32 | 705.759809017
No | Yes | 32 | 706.333632708
Yes | No | 32 | 707.185312033
No | No | 32 | 706.220334291 710.591788054
Yes | Yes | 1024* | 726.16797924
No | Yes | 1024* | 719.755843639
Yes | No | 1024* | 728.511576176
No | No | 1024* | 733.085124731

\* I had to disable `ProgressBar` in the experiments with batch size 1024 because of a `ValueError('Value out of range')` error. This might cause a slight difference in the timings, but shouldn't be significant. If you want I can run again the experiments with batch size 32, without `ProgressBar` this time.

**Comment by vdumoulin** [comment]
@fvisin thanks a lot! Would you mind posting the script you used as a gist for other people to see?

Judging from these results, it looks like we're not gaining much by supporting slices. My recommendation would be _not_ to support slices, unless someone makes a strong and well-documented case for it.

@dwf @bartvm @rizar @dmitriy-serdyuk I'd like to have your opinion as well. Do you think these results are conclusive enough? Are there some corner cases you'd like to see explored further?

**Comment by fvisin** [comment]
Here is the gist: https://gist.github.com/fvisin/fdafb338483a07fcf34f

**Update:** I noticed that the new benchmarks were not always consistent with the previous ones. I decided to run again the `No | No | 32 | model` and it took 710.591788054 (vs 706.220334291). This suggests that the benchmarks I ran might not be very reliable. So far I ran the benchmarks on my machine, but it is being used by other people and that might affect the results. 

I am running again the whole suite twice on a different machine (Sencha) that is not being used by anyone else. I will report the results.

**Comment by bartvm** [comment]
Those benchmarks don't make much sense to me... Here's my attempt to show that slices are worth supporting:

```python
import h5py
import numpy as np

# Create a dataset
f = h5py.File("benchmark.hdf5", "w")
dset = f.create_dataset("benchmark", (100000, 300), dtype='f')
dset[...] = np.random.randn(100000, 300)
f.close()

# Open the dataset
f = h5py.File("benchmark.hdf5", "r")

# Functions that return random slices/elements (random to prevent caching)
def slice(f):
    i = np.random.randint(100000 - 5000)
    return f['benchmark'][i:i + 5000]

def list(f):
    i = np.random.randint(100000 - 5000)
    return f['benchmark'][range(i, i + 5000)]
```
The results are then
```
%timeit slice(f)
The slowest run took 6.19 times longer than the fastest. This could mean that an intermediate result is being cached
1000 loops, best of 3: 1 ms per loop

%timeit list(f)
10 loops, best of 3: 43.7 ms per loop
```

Note that `%timeit range(1000, 6000)` takes only 30.5 µs, so the difference isn't due to the time it takes `range` to construct the list.

Clearly a significant difference, but my guess is that it mostly shows for large selections, which is why it doesn't show in @fvisin's benchmark. Behind the scenes, slices are directly passed to `h5py` as hyperslabs, while lists are handled with a Python loop, which includes [calls to `sorted`](https://github.com/h5py/h5py/blob/master/h5py/_hl/selections.py#L356), making things significantly slower.

**Comment by vdumoulin** [comment]
@bartvm while it's true that we can show how using lists is slower than using slices with your approach, I think we should also demonstrate that it constitutes a bottleneck in a typical machine learning use case, which is why @fvisin's approach is relevant, IMHO.

**Comment by fvisin** [comment]
Here are the new benchmark's results.

Slice | In memory |  Batch size | Time for 2 epochs (CPU) | Time for 2 epochs (CPU)
----- | --------- | ---------- | ----------------- | ----------------- 
Yes | Yes | 32 | 1266.82690668 | 1265.05306005
No | Yes | 32 | 1265.68995142 | 1267.14230585      
Yes | No | 32 | 1265.30398893 | 1265.17419457
No | No | 32 | 1265.98192334 | 1265.89525366
Yes | Yes | 1024* | 1273.51299691 | 1270.86233401
No | Yes | 1024* | 1270.44780827 | 1271.19044209
Yes | No | 1024* | 1270.53097796 | 1270.27393913
No | No | 1024* | 1271.62252188 | 1270.77412081

**Comment by fvisin** [comment]
@bartvm @vdumoulin For what concerns the meaningfulness of the tests, one is a unit test and the other an integration test. They are both meaningful, even if with different scopes.
I'll leave it to you to interpret the results, as I am not familiar with fuel's implementation.

**Comment by bartvm** [comment]
I agree that both are useful, but a 50x slowdown is hard to ignore given a single "typical use case". There are a myriad of use cases, and some of those could feasibly include a case where computation isn't much of a bottleneck (e.g. logistic regression) but where we want to use (very) large batches. In @fvisin's test there are so many confounding factors that it's impossible to attribute the difference of 0.1-0.2% to anything but noise.

**Comment by vdumoulin** [comment]
It's hard to design meaningful benchmarks unless we have a clear idea of what we're aiming for, and I realize that we haven't clarified that yet.

What are the principal use cases we'd wish to support (e.g. large-scale experiments vs small-scale experiments, complex models vs simple models)? At which point do we consider that we're facing a significant slowdown?

Unless we reach a consensus on these questions, I'm afraid we'll have a hard time finding the right benchmark to settle the *slice vs list* question.

**Comment by dwf** [comment]
@bartvm Would you be convinced by something like, large batch size (maybe each of 1024, 2048, 4096) logistic regression on SVHN? That gets us out of the territory of < 200 MB toy datasets and onto something a little more "2015-scale". But I don't want to send anyone off on this kind of errand if the results aren't going to convince anybody of anything and lead to a conclusive decision on the path forward.

I suggest we pair it down to nothing but Blocks and Fuel with no extensions or preprocessing and a very small learning rate for like 10 epochs or something.

**Comment by bartvm** [comment]
If there's no slow down on logistics regression with a 4096-batch I think you could argue that the benefits of supporting slices are not worth the engineering effort. (It would good to have a timing of the data reading part separately from the entire training though, which `BLOCKS_PROFILE=true` should give you.)

**Comment by vdumoulin** [comment]
Here are the results (the script used for benchmarking is available [here](https://gist.github.com/vdumoulin/57f8b3eb841ef9fb76ae)):

### Setup

* Logistic regression
* SVHN (format 2), training set
* SGD with fixed learning rate (1e-5)
* 10 epochs
* Batch size of 4096

### Results

Slice | In memory | Sort indices* | 3 loops, best of 3 | % time reading data (single run)
----- | --------- | ------------ | ------------------ | --------------------------------
No | No | Yes | 35.3 | 59.63%
No | No | No | 34.4 | 58.14%
Yes | No | N.A. | 16.5 | 4.94%
No | Yes | Yes | 15.7 | 5.01%
No | Yes | No | 15.9 | 5.07%
Yes | Yes | N.A. | 15.4 | 0.06%

*_Sort indices_ means that `H5PYDataset` explicitly sorts the list of indices it receives as a request.

**Comment by vdumoulin** [comment]
Pinging @dwf, @bartvm and @rizar. I'd be interested in having your opinion in the light of the most recent benchmarks.

**Comment by dwf** [comment]
The results seem to suggest that it's worth taking the engineering hit and
supporting slices. If you're going to be ever spending >50% of your time
reading data...

**Comment by bartvm** [comment]
Agreed with @dwf, especially considering the engineering effort should be relatively small

**Comment by rizar** [comment]
+1, apparently we have to support slices.

I think however that the current code for supporting slices is quite
complex and significantly affects the readability. It seems that we can do
better by factoring out a `Subset` class, which instance would incapsulate
either slice or list. Such a class can support operations like
concatenation and translation and implement them in the most efficient way
possible, that is returning slices whenever possible. Such a class could be
even reusable outside of `H5PyDataset`.

**Comment by vdumoulin** [comment]
@rizar I like this idea, I'll start working on that. Thanks for your input, guys!
